### PR TITLE
feat: Add template node to pod name. Fixes #1319

### LINKED ISSUE #1319: Add taskname to the workflow pod's name
**Is this a BUG REPORT or FEATURE REQUEST?**:
Feature Request

**What happened**:
Currently, running a workflow creates a pod for each task, and each of those pods gets a unique name associated with the workflow. So if I run a workflow which has 3 steps, then I run `kubectl get pods`, I get something like this:

```
NAME                                  READY     STATUS              RESTARTS   AGE
wf-install-1938236639                 0/1       Completed           0          16s
wf-install-3464818                    0/2       Completed           0          23s
wf-install-3672310916                 0/2       Completed           0          23s
wf-install-376633117                  0/2       Completed           0          23s
```

**What you expected to happen**:
This is pretty annoying, because it makes it very hard to determine which pod is doing which task. I propose adding the task name to the pod name, such that the previous step looks something like the following:
```
NAME                                  READY     STATUS              RESTARTS   AGE
wf-install-1938236639                 0/1       Completed           0          16s
wf-install-db-init-3464818            0/2       Completed           0          23s
wf-install-setup-host-3672310916      0/2       Completed           0          23s
wf-install-send-email-376633117       0/2       Completed           0          23s
```

### PR DESCRIPTION
**Pull request by JPZ13**
This PR is a continuation of Alex's PoC  #6647 that fixes #1319. It:
- Sets pod names to a concatenation of the wf name, template name, and hash
- Allows for an override of the pod name to the node id if the env `USE_LEGACY_POD_NAMES` is set to true
- Stores the node id as an annotation on the pod

Sample terminal output:
```
$ kubectl get pods -n argo
NAME                                   READY   STATUS      RESTARTS   AGE
minio-786d5cbc44-c9ppg                 1/1     Running     7          19h
retry-test-p7jzr-whalesay-2308805457   0/2     Completed   0          13h
retry-test-p7jzr-whalesay-2378053249   0/2     Completed   0          13h
```

### CODE REVIEW

File: docs/environment-variables.md [hunk 1]
```diff
@@ -32,6 +32,7 @@ Note that these environment variables may be removed at any time.
 | `RETRY_BACKOFF_STEPS` | `int` | `5` | The retry backoff steps when retrying API calls. |
 | `RETRY_HOST_NAME_LABEL_KEY` | `string` | `kubernetes.io/hostname` | The label key for host name used when retrying templates. |
 | `TRANSIENT_ERROR_PATTERN` | `string` | `""` | The regular expression that represents additional patterns for transient errors. |
+| `USE_LEGACY_POD_NAMES` | `bool` | `false` | Whether to have pod names contain the template name (false) or be the node id (true). |
```
Comment (alexec) [inline docs/environment-variables.md:35]: how about `POD_NAMES=v1` default is `v2`

File: workflow/controller/node_counters_test.go [hunk 2]
```diff
@@ -147,4 +147,17 @@ func TestCounters(t *testing.T) {
 	assert.Equal(t, int64(2), woc.getActivePods("2"))
 	assert.Equal(t, int64(2), woc.getActiveChildren("2"))
 	assert.Equal(t, int64(2), woc.getUnsuccessfulChildren("2"))
+
+	testNodePodExists(t, woc)
+}
+
+func testNodePodExists(t *testing.T, woc *wfOperationCtx) {
```
Comment (alexec) [inline workflow/controller/node_counters_test.go:154]: should be `Test` uppercase T

File: workflow/controller/workflowpod.go [hunk 3]
```diff
@@ -205,17 +204,17 @@ func (woc *wfOperationCtx) createWorkflowPod(ctx context.Context, nodeName strin
...
-			Name:      nodeID,
+			Name:      util.PodName(woc.wf.Name, nodeName, tmpl.Name, nodeID),
 			Namespace: woc.wf.ObjectMeta.Namespace,
 			Labels: map[string]string{
 				common.LabelKeyWorkflow:  woc.wf.ObjectMeta.Name, // Allows filtering by pods related to specific workflow
 				common.LabelKeyCompleted: "false",                // Allows filtering by incomplete workflow pods
 			},
 			Annotations: map[string]string{
 				common.AnnotationKeyNodeName: nodeName,
+				common.AnnotationKeyNodeID:   nodeID, // TODO - test annotation is applied
```
Comment (alexec) [inline workflow/controller/workflowpod.go:217]: TODO

File: workflow/util/util.go [hunk 4]
```diff
@@ -1054,3 +1061,31 @@ func GetNodeType(tmpl *wfv1.Template) wfv1.NodeType {
 	}
 	return ""
 }
+
+// PodName return a deterministic pod name
+func PodName(workflowName, nodeName, templateName, nodeID string) string {
```
Comment (alexec) [inline workflow/util/util.go:1066]: please put this func into own file

File: workflow/util/util_test.go [hunk 5]
```diff
@@ -839,3 +840,34 @@ func TestToUnstructured(t *testing.T) {
 		assert.Equal(t, workflow.Version, gv.Version)
 	}
 }
+
+func TestPodName(t *testing.T) {
```
Comment (alexec) [inline workflow/util/util_test.go:844]: own file

**Review by alexec — CHANGES_REQUESTED** [review]
some small changes

**Comment by alexec** [comment]
LGTM. Passing to @sarabala1979 for him to review.

**Review by alexec — APPROVED** [review]

**Comment by alexec** [comment]
@sarabala1979 I'm going to assume you're lack of reply means you are not able to review this. I'll merge.

**Comment by vivekthangathurai** [comment]
I am using v3.2.8 and not seeing this behaviour.

**Pull request merged**